fabric-brewing-v1/src/main/java/net/fabricmc/fabric/impl/brewing/PotionIngredient.java

```diff
@@ -0,0 +1,110 @@
+import net.minecraft.util.PacketByteBuf;
+import net.minecraft.util.registry.Registry;
+
+public class PotionIngredient {
+    private final Potion potion;
+    private final Ingredient delegate;
+    private final Type type;
+
+    private PotionIngredient(Potion potion, Type type) {
+        this.potion = potion;
+        this.type = type != null ? type : Type.NORMAL;
+        this.delegate = null;
+    }
+
+    private PotionIngredient(Ingredient delegate) {
+        this.potion = null;
+        this.type = null;
+        this.delegate = delegate;
+    }
+
+    public boolean isIngredient() { return delegate != null; }
+    public boolean isPotion() { return potion != null; }
+
+    public Ingredient asIngredient() {
+        return this.isIngredient() ? delegate
+            : Ingredient.ofStacks(PotionUtil.setPotion(new ItemStack(Items.POTION), potion)); }
+    public Potion asPotion() { return potion; }
+    public Type getPotionType() { return type; }
+
+    public boolean test(ItemStack stack) {
+        return this.isIngredient() ? delegate.test(stack) :
+            PotionUtil.getPotion(stack) == potion && stack.getItem() == getPotionType().item; }
+
+	public static PotionIngredient fromJson(JsonElement json) { return fromJson(json, false); }
+    public static PotionIngredient fromJson(JsonElement json, boolean forceSingle) {
+        if(json != null && !json.isJsonNull()) {
+            if(json.isJsonObject()) {
+                JsonObject obj = json.getAsJsonObject();
+                if(forceSingle && JsonHelper.hasString(obj, "tag"))
+                	throw new JsonSyntaxException("Cannot use a tag for recipe output");
+
+                return JsonHelper.hasString(obj, "potion")
+                    ? new PotionIngredient(Registry.POTION.get(new Identifier(JsonHelper.getString(obj, "potion"))),
+                        Type.fromString(JsonHelper.getString(obj, "type", "normal")))
+                    : new PotionIngredient(Ingredient.fromJson(obj));
+            } else
+                throw new JsonSyntaxException("Expected item to be object");
+        } else
+            throw new JsonSyntaxException("Item cannot be null");
+    }
+
+    public static PotionIngredient fromPacket(PacketByteBuf buf) {
+        boolean isPotion = buf.readBoolean();
+        if(!isPotion) return new PotionIngredient(Ingredient.fromPacket(buf));
+        return new PotionIngredient(Registry.POTION.get(new Identifier(buf.readString())),
+            Type.fromString(buf.readString()));
+    }
+
+    public void write(PacketByteBuf buf) {
+        buf.writeBoolean(this.isPotion());
+        if(!this.isPotion()) this.delegate.write(buf);
+        else {
+            buf.writeString(Registry.POTION.getId(this.potion).toString());
+            buf.writeString(type.toString().toLowerCase());
+        }
+    }
+
+    public enum Type {
+        NORMAL(Items.POTION),
+        SPLASH(Items.SPLASH_POTION),
+        LINGERING(Items.LINGERING_POTION);
```

Comment:
Don't use an enum in an API when you or another mod may add more entries to it in the future.

Reply:
Noted, I'll see about changing that.